File: backend/src/main/java/com/example/realestate/model/ProjectMilestone.java
package com.example.realestate.model;

import com.fasterxml.jackson.annotation.JsonFormat;
import jakarta.persistence.*;

import java.time.LocalDate;

@Entity
@Table(name = "project_milestones")
public class ProjectMilestone {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "project_id", nullable = false)
    private Long projectId;

    @Column(name = "investor_id", nullable = false)
    private Long investorId;

    @Column(nullable = false)
    private String milestone;

    @Column(columnDefinition = "text")
    private String description;

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
    private LocalDate date;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private MilestoneStatus status; // e.g., PENDING, COMPLETED

    public ProjectMilestone() {
    }

    public ProjectMilestone(Long projectId, Long investorId, String milestone, String description, LocalDate date, MilestoneStatus status) {
        this.projectId = projectId;
        this.investorId = investorId;
        this.milestone = milestone;
        this.description = description;
        this.date = date;
        this.status = status;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Long getProjectId() {
        return projectId;
    }

    public void setProjectId(Long projectId) {
        this.projectId = projectId;
    }

    public Long getInvestorId() {
        return investorId;
    }

    public void setInvestorId(Long investorId) {
        this.investorId = investorId;
    }

    public String getMilestone() {
        return milestone;
    }

    public void setMilestone(String milestone) {
        this.milestone = milestone;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public LocalDate getDate() {
        return date;
    }

    public void setDate(LocalDate date) {
        this.date = date;
    }

    public MilestoneStatus getStatus() {
        return status;
    }

    public void setStatus(MilestoneStatus status) {
        this.status = status;
    }
}

File: lib/models/project_milestone.dart
class ProjectMilestone {
  final int? id;
  final int projectId;
  final int investorId;
  final String milestone;
  final String? description;
  final String? date; // yyyy-MM-dd
  final String status; // PENDING, COMPLETED

  ProjectMilestone({
    this.id,
    required this.projectId,
    required this.investorId,
    required this.milestone,
    this.description,
    this.date,
    this.status = 'PENDING',
  });

  factory ProjectMilestone.fromJson(Map<String, dynamic> json) {
    return ProjectMilestone(
      id: json['id'],
      projectId: json['projectId'],
      investorId: json['investorId'],
      milestone: json['milestone'] ?? '',
      description: json['description'],
      date: json['date'],
      status: json['status'] ?? 'PENDING',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'projectId': projectId,
      'investorId': investorId,
      'milestone': milestone,
      'description': description,
      'date': date,
      'status': status,
    };
  }
}
